### PR TITLE
rbd-mirror: initial snapshot-based mirroring bootstrap logic

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2020-01-31_12:27:51-rbd-wip-jd-testing-distro-basic-smithi/

File: src/librbd/api/Mirror.cc [hunk 1]
```diff
@@ -1816,9 +1842,30 @@ int Mirror<I>::image_snapshot_create(I *ictx, uint64_t *snap_id) {
...
+    return -EINVAL;
+  } else if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror image" << dendl;
+    return r;
+  }
+
+  if (mirror_image.mode != cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT ||
+      mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+    lderr(cct) << "snapshot based mirroring is not enabled" << dendl;
+    return r;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: @dillaman want to return `-EINVAL` here?

Comment (dillaman) [reply src/librbd/api/Mirror.cc]: Yeah, I fixed it in the next PR [1] but I can cherry-pick over here.

[1] https://github.com/ceph/ceph/pull/33097/commits/9538bd8a11a986469a86e2fbd64858e95ddd1b77

**Comment by trociny** [comment]
@dillaman the python `test_rbd.TestMirroring.test_mirror_image_create_snapshot` needs update. Previously it failed due to not correct error code from image_snapshot_create API, and after the fix applied it still fails: it does not take into account the mirror snapshot that is created on image mirror enable:
```
FAIL: test_rbd.TestMirroring.test_mirror_image_create_snapshot
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/mgolub/ceph/ceph.ci/src/test/pybind/test_rbd.py", line 2090, in test_mirror_image_create_snapshot
    eq(1, len(snaps))
AssertionError: 1 != 2
```

**Comment by dillaman** [comment]
> @dillaman the python test_rbd.TestMirroring.test_mirror_image_create_snapshot needs update. 

Sorry, I had fixed that in a later commit as well. Now pushed to this PR.

**Comment by dillaman** [comment]
jenkins test make check

**Pull request merged**